### DISCUSSION TITLE
Next.js RSC vs Static Exports Mode Best For PWA?

### DISCUSSION THREAD

**Discussion by bestickley** [serwist/serwist#140] [Q&A] [answered]
Hi, I need to build a offline first PWA. I'd prefer to use Next.js 14 App Router with all it's (server) features, but I'm wondering if it that will make building an offline-first / PWA difficult. Should I resort to using Next.js static exports mode?

**Comment by DuCanhGH** [comment, marked as answer]
@bestickley 

> RSC vs Static Exports

They are not mutually exclusive, fyi.

Well, if you keep your pages dynamically generated, just don't precache them and you should be fine :) `@serwist/next` serves them with the `NetworkFirst` strategy out of the box. If you want to go down the offline-first route, it's recommended that you statically generate your application and keep dynamic content on the client (I'm not sure about PPR, please do your research on how it works, thanks!). To precache the pages, use the `additionalPrecacheEntries` option.

**Reply by bestickley** [reply]
@DuCanhGH, thank you for quick response! I thought they were mutually exclusive because when you set: `output: 'export',` it applies to the whole app, right? Are you referring to the fact that RSC can pre-render/build?

I guess I don't want to go down the `offline-first` route. `NetworkFirst` sounds like a good option with RSC.

It seems like if I want to use RSC, I should make it so that when user initially connects to home page (online), it then (in background) fetches and caches all other pages so that when user goes offline and then tries to visit new page other than home, it will be cached and available, right? Do I need `additionalPrecacheEntries` for that, or will `precacheEntries: self.__SW_MANIFEST,` work? If there are docs on this, please forgive me and point me in that direction.

**Reply by DuCanhGH** [reply]
@bestickley

> Are you referring to the fact that RSC can pre-render/build?

Yeah...

> I guess I don't want to go down the offline-first route. NetworkFirst sounds like a good option with RSC.

Well, you may precache some statically generated pages :) That may cause version mismatches between statically generated pages and dynamically generated ones though, but I'm not sure how Next.js handles that.

> It seems like if I want to use RSC, I should make it so that when user initially connects to home page (online), it then (in background) fetches and caches all other pages so that when user goes offline and then tries to visit new page other than home, it will be cached and available, right? Do I need additionalPrecacheEntries for that, or will precacheEntries: self.__SW_MANIFEST, work? If there are docs on this, please forgive me and point me in that direction.

Again, ~~RSC~~ server features. RSC is... well, not that great of a name, but you may want to think that the prerendering process is simply a server preprocessing page requests into HTML responses that will be reused later.

But no, please don't precache dynamically generated pages. This causes them to be served from the precache until they are removed from your precache list. You should use `window.serwist.messageSW({ type: "CACHE_URLS", urlsToCache: ["/a", "/b"] });` instead. If you use Serwist's service worker API, each URL in `urlsToCache` will be handled by a matching runtime caching strategy. Otherwise, you will need to handle this message yourself.

**Reply by bestickley** [reply]
Got it. I will dive deeper into this. Thank you so much!

**Reply by bestickley** [reply]
@DuCanhGH, re:
> But no, please don't precache dynamically generated pages. This causes them to be served from the precache until they are removed from your precache list.

But what if I use NetworkFirst strategy? Wouldn't that make it so even if I cache dynamic route, that upon subsequent load of that dynamic page it would attempt to get new one from network, and only fallback to cache if network fails? Or does `additionalPrecacheEntries` specifically not work like that?

Ideally, upon initial load of first page of app, I'd like to cache all routes so that if user then goes offline, they can visit all pages of app. I understand that those pages will only be as up to date as when the user was last offline.

**Reply by DuCanhGH** [reply]
@bestickley 

Precached URLs are NOT handled by any of your runtime caching strategies. They have their own caching strategy, which is always tested first, meaning that it always matches those URLs.

> Ideally, upon initial load of first page of app, I'd like to cache all routes so that if user then goes offline, they can visit all pages of app. I understand that those pages will only be as up to date as when the user was last offline.

I will repeat this again: please use `window.serwist.messageSW({ type: "CACHE_URLS", urlsToCache: [] })`. That will work the way you expect it to. It forces Serwist to cache the URLs without them being visited.

One big issue is that Serwist never knows what URLs your Next.js application will have. If you want to cache all routes, you have to feed it all of them.

**Reply by bestickley** [reply]
@DuCanhGH, got it. I guess I didn't understand "precached" has a special meaning. I will use that API. Thank you, again!

**Reply by DuCanhGH** [reply]
@bestickley on precaching and runtime caching:

- https://serwist.pages.dev/docs/serwist/guide/precaching

- https://serwist.pages.dev/docs/serwist/runtime-caching

**Reply by bestickley** [reply]
@DuCanhGH, perfect. Was looking for this. Thank you!

**Comment by paul-stinchcombe** [comment]
I build a PWA with NextJS 14 without issues.

**Reply by bestickley** [reply]
@paul-stinchcombe, thanks for the response. Would you mind sharing your next.config.js? Do you precache all routes upon initial load so that the entire app is available offline? Or do you only cache as they're visited so that only if you've visited a route while online it's available offline?
